priv/perf/apps/load_testing/lib/utils/faucet.ex

```diff
@@ -0,0 +1,136 @@
+  @moduledoc false
+
+  require Logger
+
+  use GenServer
+
+  alias ExPlasma.Encoding
+  alias ExPlasma.Utxo
+  alias LoadTesting.Utils.Account
+  alias LoadTesting.Utils.ChildChain
+  alias LoadTesting.Utils.Deposit
+  alias LoadTesting.Utils.Ethereum, as: Eth
+
+  @eth <<0::160>>
+  @fund_child_chain_account_retries 100
+
+  defstruct [:account, :fee, utxos: %{}]
+
+  def start_link(_) do
+    GenServer.start_link(__MODULE__, nil, name: __MODULE__)
+  end
+
+  def fund_child_chain_account(account, amount, token) do
+    GenServer.call(__MODULE__, {:fund_child_chain, account, amount, token}, :infinity)
+  end
+
+  def get_faucet() do
+    GenServer.call(__MODULE__, :get_faucet, :infinity)
+  end
+
+  def init(_) do
+    opts = fetch_default_opts()
+    fee_wei = Keyword.fetch!(opts, :fee_wei)
+
+    {faucet, eth_utxo} = get_funded_faucet_account(opts)
+    Logger.debug("Using faucet: #{Encoding.to_hex(faucet.addr)}")
+
+    state = struct!(__MODULE__, account: faucet, fee: fee_wei, utxos: %{@eth => eth_utxo})
+    {:ok, state}
+  end
+
+  def handle_call(:get_faucet, _from, %__MODULE__{account: faucet} = state) do
+    {:reply, {:ok, faucet}, state}
+  end
+
+  def handle_call(
+        {:fund_child_chain, account, amount, @eth = token},
+        _from,
+        %__MODULE__{account: faucet} = state
+      ) do
+    {utxo, utxo_amount} = state.utxos[token]
+    change = utxo_amount - amount - state.fee
+    if change < 0, do: raise({:error, :change_below_zero})
+
+    Logger.debug("Funding user: #{Encoding.to_hex(account.addr)} with UTXO: #{Utxo.pos(utxo)}")
+
+    outputs = [
+      %Utxo{amount: change, currency: token, owner: faucet.addr},
+      %Utxo{amount: amount, currency: token, owner: account.addr}
+    ]
+
+    {:ok, blknum, txindex} = ChildChain.submit_tx([utxo], outputs, [faucet], @fund_child_chain_account_retries)
+
+    {:ok, change_utxo} = Utxo.new(%{blknum: blknum, txindex: txindex, oindex: 0})
+    {:ok, user_utxo} = Utxo.new(%{blknum: blknum, txindex: txindex, oindex: 1})
+
+    updated_state = Map.put(state, :utxos, %{state.utxos | token => {change_utxo, change}})
+
+    {:reply, {:ok, {Utxo.pos(user_utxo), amount}}, updated_state}
+  end
+
+  defp get_funded_faucet_account(opts) do
+    faucet =
+      case Keyword.fetch(opts, :faucet) do
+        {:ok, faucet} ->
+          faucet
+
+        :error ->
+          Logger.debug("Generating a faucet")
+          {:ok, faucet} = create_funded_faucet(opts)
+          faucet
+      end
+
+    deposit_amount = Keyword.fetch!(opts, :faucet_deposit_wei)
+    deposit_finality_margin = Keyword.fetch!(opts, :deposit_finality_margin)
+
+    {:ok, deposit_utxo} = Deposit.deposit_from(faucet, deposit_amount, @eth, deposit_finality_margin)
+
+    {faucet, {deposit_utxo, deposit_amount}}
+  end
+
+  defp create_funded_faucet(opts) do
+    # Sends a transaction to a local instance of geth.
+    # NOTE: Assumes existence of an unlocked account managed by Ethereum client.
```

Comment:
what does this sentence mean? does it mean it doesn't work with infura?

Reply:
it indeed means that, but is only used in case of test environment. On `infura` we have a founded faucet, but in case of tests we just create a faucet ad hoc, funding it from unlocked accounts, that `geth` dev instance provides.